File: backend/app/schemas/identidade.py
"""Schemas Pydantic v2 — GET /identidade/re/{re} (Bloco A2).

Endpoint compartilhado de identidade por RE, pra pré-preenchimento em
qualquer módulo (portaria, ocorrência, pré-ocorrência) sem duplicar a
busca em public.funcionario/public.motorista (ver app/services/identidade.py,
que já resolve as duas tabelas — §5.3).

⛔ Nunca cpf/rg/cnh/telefone nem histórico — este endpoint só confirma
visualmente quem é o RE e o que a portaria precisa pra acelerar a entrada.
O autopreenchimento mais rico da Ocorrência (que já traz documentos)
continua em GET /ocorrencias/autopreencher/pessoa, não aqui.
"""
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel


class VeiculoParticularIdentidade(BaseModel):
    id: UUID
    placa: str
    marca_modelo: Optional[str] = None
    cor: Optional[str] = None


class IdentidadeReResponse(BaseModel):
    # ⛔ Nunca 404 — RE inexistente é resultado válido (regra número um,
    # mesma convenção de ResolverReResponse em schemas/portaria.py).
    encontrado: bool = False
    id: Optional[UUID] = None
    nome: Optional[str] = None
    origem: Optional[Literal["FUNCIONARIO", "MOTORISTA"]] = None
    ativo: Optional[bool] = None
    # Só preenchido quando origem=FUNCIONARIO — motorista (legado) não tem
    # vínculo de função no cadastro central.
    funcoes: list[str] = []
    # PARTICULAR de portaria.veiculo, dono = esta pessoa, ativo=true. Lista
    # porque a mesma pessoa pode ter mais de um veículo particular
    # cadastrado.
    veiculo_particular: list[VeiculoParticularIdentidade] = []

File: frontend-v3/assets/js/identidade.js
/*
 * identidade.js — busca por RE compartilhada entre módulos (Bloco A2)
 * -----------------------------------------------------------------------
 * Consome GET /identidade/re/{re} (nome, origem, funções, veículo
 * particular — ⛔ nunca cpf/rg/cnh/telefone). Um RE, um lugar de busca:
 * nenhum módulo deveria reimplementar "digitar RE, ver nome" com a própria
 * chamada — quem já tem autopreenchimento mais rico e próprio (ver
 * ocorrencia.form.js::autopreencherPessoa, que também traz documentos via
 * /ocorrencias/autopreencher/pessoa) continua com o endpoint dele.
 *
 * ⚠️ Exige sessão autenticada (JWT via api.js) — não funciona nas páginas
 * públicas por token (pre-ocorrencia.page.js), que não têm apiGet.
 */
import { apiGet, ApiError } from './api.js';

export async function buscarPorRe(re) {
    const limpo = (re || '').trim();
    if (!limpo) return null;
    return apiGet(`/identidade/re/${encodeURIComponent(limpo)}`);
}

/**
 * Liga a busca por RE a um input: dispara no blur, preenche só campo
 * vazio (nunca sobrescreve o que a pessoa digitou) e deixa tudo editável
 * — mesma regra do autopreenchimento da Ocorrência (10/08/2026).
 *
 * @param {HTMLInputElement} inputRe
 * @param {object} [opcoes]
 * @param {(dados: object) => void} [opcoes.onEncontrado] — dados = IdentidadeReResponse
 * @param {() => void} [opcoes.onNaoEncontrado]
 * @param {Record<string, HTMLInputElement>} [opcoes.campos] — nome do campo
 *   da resposta (hoje só faz sentido pra "nome") -> input a preencher se
 *   estiver vazio. Campos compostos (funcoes, veiculo_particular) são
 *   responsabilidade do onEncontrado, não deste mapeamento.
 * @param {(re: string) => Promise<object|null>} [opcoes.buscar] — sobrescreve
 *   a busca padrão; permite reusar este fiação (blur + "só campo vazio")
 *   com um endpoint mais rico (caso da Ocorrência).
 * @param {number} [opcoes.minimoDigitos]
 */
export function ligarAutopreenchimentoRe(inputRe, opcoes = {}) {
    const {
        onEncontrado, onNaoEncontrado, campos = {}, buscar = buscarPorRe, minimoDigitos = 3,
    } = opcoes;

    async function disparar() {
        const re = inputRe.value.trim();
        if (re.length < minimoDigitos) return;
        let dados;
        try {
            dados = await buscar(re);
        } catch (err) {
            if (err instanceof ApiError && err.status === 401) return;
            console.error('[identidade] erro ao buscar por RE:', err);
            return;
        }
        if (!dados || !dados.encontrado) {
            if (onNaoEncontrado) onNaoEncontrado();
            return;
        }
        for (const [campo, el] of Object.entries(campos)) {
            const valor = dados[campo];
            if (el && !el.value.trim() && typeof valor === 'string' && valor) {
                el.value = valor;
            }
        }
        if (onEncontrado) onEncontrado(dados);
    }

    inputRe.addEventListener('blur', disparar);
}
